{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c0356d10-947d-4b71-bdc8-77317d6cfac2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MEGADNA(\n",
       "  (start_tokens): ParameterList(\n",
       "      (0): Parameter containing: [torch.float32 of size 512 (GPU 0)]\n",
       "      (1): Parameter containing: [torch.float32 of size 256 (GPU 0)]\n",
       "      (2): Parameter containing: [torch.float32 of size 196 (GPU 0)]\n",
       "  )\n",
       "  (token_embs): ModuleList(\n",
       "    (0): Embedding(6, 196)\n",
       "    (1): Sequential(\n",
       "      (0): Embedding(6, 196)\n",
       "      (1): Rearrange('... r d -> ... (r d)')\n",
       "      (2): LayerNorm((3136,), eps=1e-05, elementwise_affine=True)\n",
       "      (3): Linear(in_features=3136, out_features=256, bias=True)\n",
       "      (4): LayerNorm((256,), eps=1e-05, elementwise_affine=True)\n",
       "    )\n",
       "    (2): Sequential(\n",
       "      (0): Embedding(6, 196)\n",
       "      (1): Rearrange('... r d -> ... (r d)')\n",
       "      (2): LayerNorm((200704,), eps=1e-05, elementwise_affine=True)\n",
       "      (3): Linear(in_features=200704, out_features=512, bias=True)\n",
       "      (4): LayerNorm((512,), eps=1e-05, elementwise_affine=True)\n",
       "    )\n",
       "  )\n",
       "  (transformers): ModuleList(\n",
       "    (0): Transformer(\n",
       "      (layers): ModuleList(\n",
       "        (0-7): 8 x ModuleList(\n",
       "          (0): Attention(\n",
       "            (attend): Attend(\n",
       "              (attn_dropout): Dropout(p=0.0, inplace=False)\n",
       "            )\n",
       "            (dropout): Dropout(p=0.0, inplace=False)\n",
       "            (norm): RMSNorm()\n",
       "            (to_q): Linear(in_features=512, out_features=512, bias=False)\n",
       "            (to_kv): Linear(in_features=512, out_features=128, bias=False)\n",
       "            (to_out): Linear(in_features=512, out_features=512, bias=False)\n",
       "          )\n",
       "          (1): Sequential(\n",
       "            (0): RMSNorm()\n",
       "            (1): Linear(in_features=512, out_features=2048, bias=True)\n",
       "            (2): GELU(approximate='none')\n",
       "            (3): Dropout(p=0.0, inplace=False)\n",
       "            (4): Linear(in_features=2048, out_features=512, bias=True)\n",
       "          )\n",
       "        )\n",
       "      )\n",
       "      (norm): RMSNorm()\n",
       "    )\n",
       "    (1): Transformer(\n",
       "      (layers): ModuleList(\n",
       "        (0-7): 8 x ModuleList(\n",
       "          (0): Attention(\n",
       "            (attend): Attend(\n",
       "              (attn_dropout): Dropout(p=0.0, inplace=False)\n",
       "            )\n",
       "            (dropout): Dropout(p=0.0, inplace=False)\n",
       "            (norm): RMSNorm()\n",
       "            (to_q): Linear(in_features=256, out_features=512, bias=False)\n",
       "            (to_kv): Linear(in_features=256, out_features=128, bias=False)\n",
       "            (to_out): Linear(in_features=512, out_features=256, bias=False)\n",
       "          )\n",
       "          (1): Sequential(\n",
       "            (0): RMSNorm()\n",
       "            (1): Linear(in_features=256, out_features=1024, bias=True)\n",
       "            (2): GELU(approximate='none')\n",
       "            (3): Dropout(p=0.0, inplace=False)\n",
       "            (4): Linear(in_features=1024, out_features=256, bias=True)\n",
       "          )\n",
       "        )\n",
       "      )\n",
       "      (norm): RMSNorm()\n",
       "    )\n",
       "    (2): Transformer(\n",
       "      (layers): ModuleList(\n",
       "        (0-7): 8 x ModuleList(\n",
       "          (0): Attention(\n",
       "            (attend): Attend(\n",
       "              (attn_dropout): Dropout(p=0.0, inplace=False)\n",
       "            )\n",
       "            (dropout): Dropout(p=0.0, inplace=False)\n",
       "            (norm): RMSNorm()\n",
       "            (to_q): Linear(in_features=196, out_features=512, bias=False)\n",
       "            (to_kv): Linear(in_features=196, out_features=128, bias=False)\n",
       "            (to_out): Linear(in_features=512, out_features=196, bias=False)\n",
       "          )\n",
       "          (1): Sequential(\n",
       "            (0): RMSNorm()\n",
       "            (1): Linear(in_features=196, out_features=784, bias=True)\n",
       "            (2): GELU(approximate='none')\n",
       "            (3): Dropout(p=0.0, inplace=False)\n",
       "            (4): Linear(in_features=784, out_features=196, bias=True)\n",
       "          )\n",
       "        )\n",
       "      )\n",
       "      (norm): RMSNorm()\n",
       "    )\n",
       "  )\n",
       "  (to_next_transformer_projections): ModuleList(\n",
       "    (0): Sequential(\n",
       "      (0): Rearrange('b ... d -> b (...) d')\n",
       "      (1): Linear(in_features=512, out_features=16384, bias=True)\n",
       "      (2): Rearrange('b m (n d) -> (b m) n d', n=64)\n",
       "    )\n",
       "    (1): Sequential(\n",
       "      (0): Rearrange('b ... d -> b (...) d')\n",
       "      (1): Linear(in_features=256, out_features=3136, bias=True)\n",
       "      (2): Rearrange('b m (n d) -> (b m) n d', n=16)\n",
       "    )\n",
       "    (2): Identity()\n",
       "  )\n",
       "  (to_logits): Linear(in_features=196, out_features=6, bias=True)\n",
       ")"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import torch\n",
    "from Bio import SeqIO\n",
    "from BCBio import GFF\n",
    "import random\n",
    "\n",
    "# load the model\n",
    "model_path = \"megaDNA_phage_145M.pt\" # model name\n",
    "device = 'cuda' if torch.cuda.is_available() else 'cpu'  # Device configuration\n",
    "\n",
    "model = torch.load(model_path, map_location=torch.device(device))\n",
    "model.eval() "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a8c0f1b",
   "metadata": {},
   "source": [
    "Please download the fasta file and gene annotation for lambda phage from https://www.ncbi.nlm.nih.gov/nuccore/NC_001416.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "48471cd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the FASTA file\n",
    "fasta_file_path = \"NC_001416.1.fasta\"\n",
    "seq_ids, sequences = [], []\n",
    "\n",
    "with open(fasta_file_path, \"r\") as fasta_file:\n",
    "    for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "        seq_ids.append(record.id)\n",
    "        sequences.append(str(record.seq))\n",
    "\n",
    "# Read the gene annotations\n",
    "gff_file_path = \"NC_001416.1.gff3\"\n",
    "limit_info = dict(gff_type=[\"CDS\"])\n",
    "\n",
    "start_position, end_position, strand_position = [], [], []\n",
    "\n",
    "with open(gff_file_path) as in_handle:\n",
    "    for rec in GFF.parse(in_handle, limit_info=limit_info):\n",
    "        start_position.extend(feature.location.start for feature in rec.features)\n",
    "        end_position.extend(feature.location.end for feature in rec.features)\n",
    "        strand_position.extend(feature.location.strand for feature in rec.features)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3e0b2f2d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tensor(1.2624, device='cuda:0')\n"
     ]
    }
   ],
   "source": [
    "nt = ['**', 'A', 'T', 'C', 'G', '#']  # Vocabulary\n",
    "seq_id = 0  # Sequence ID\n",
    "\n",
    "def encode_sequence(sequence, nt_vocab=nt):\n",
    "    \"\"\"Encode a DNA sequence to its numerical representation.\"\"\"\n",
    "    return [0] + [nt_vocab.index(nucleotide) if nucleotide in nt_vocab else 1 for nucleotide in sequence] + [5]\n",
    "\n",
    "def get_loss_for_sequence(model, sequence, device):\n",
    "    \"\"\"Get model loss for a given sequence.\"\"\"\n",
    "    input_seq = torch.tensor(sequence).unsqueeze(0).to(device)\n",
    "    with torch.no_grad():\n",
    "        loss = model(input_seq, return_value='loss')\n",
    "    return loss\n",
    "\n",
    "# Get the model loss for the WT sequence\n",
    "encoded_wt_sequence = encode_sequence(sequences[seq_id])\n",
    "wt_loss = get_loss_for_sequence(model, encoded_wt_sequence, device)\n",
    "print(wt_loss)\n",
    "\n",
    "# Get the model loss for the mutants in the start codons\n",
    "loss_start = []\n",
    "random.seed(42)\n",
    "for j, (start, end, strand) in enumerate(zip(start_position, end_position, strand_position)):\n",
    "    encoded_mutant_sequence = encode_sequence(sequences[seq_id])\n",
    "    \n",
    "    # Mutate start codon positions based on strand orientation\n",
    "    positions = range(start+1, start+4) if strand == 1 else range(end-2, end+1)\n",
    "    for i in positions:\n",
    "        encoded_mutant_sequence[i] = random.choice([1, 2, 3, 4])\n",
    "    \n",
    "    # Get model loss for mutated sequence\n",
    "    mutant_loss = get_loss_for_sequence(model, encoded_mutant_sequence, device)\n",
    "    loss_start.append(mutant_loss)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b01cb45",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (myenv)",
   "language": "python",
   "name": "transformers2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
